python_modules/dagster/dagster/_core/definitions/multi_dimensional_partitions.py

```diff
@@ -426,7 +435,11 @@ def get_cron_schedule(
         ).get_cron_schedule(minute_of_hour, hour_of_day, day_of_week, day_of_month)
 
     @property
-    def timezone(self) -> Optional[str]:
+    def fmt(self) -> str:
+        return cast(TimeWindowPartitionsDefinition, self.time_window_dimension.partitions_def).fmt
+
+    @property
+    def timezone(self) -> str:
```

Comment:
I have a similar comment about the pre-existing timezone property (which I think is just used in a couple places, where we can refactor)

In general, I think we should be consistent in the pattern that we expect callsites to have. Either we should treat a MultiPartitionsDefinition exactly like a TimeWindowPartitionsDefinition, and access the same properties (I'd rather not), or we should have a standard method to extract the time window dimension stuff from a multi partitions definition, as proposed above

note that this would somewhat obsolete `has_one_dimension_time_window_partitioning`, which I think is ok, as in a few places we call this method and then immediately do the work to parse out the TimeWindowPartitionsDefinition

Reply:
Initially my plan was to treat a MultiPartitionsDefinition like a TimeWindowPartitionsDefinition, since there likely is a maximum of 1 time dimension and all of these functions can just call the underlying `TimeWindowPartitionsDefinition`. method. 

I'm fine with refactoring this, but I'm wondering why you don't want to have a pattern like this?

Reply:
Yeah I should have explained my reasoning a bit more here. The basic idea is that in most callsites, we have an input of type `Optional[PartitionsDefinition]`. There are many categories this `partitions_def` can be in. Either:

- None
- TimeWindowPartitionsDefinition
- MultiPartitionsDefinition (exactly one time window dimension)
- MultiPartitionsDefinition (not exactly one time window dimension)
- Other PartitionsDefinition (static/dynamic)

Now, in any callsite that accepts a parameter of type PartitionsDefinition, where we want to treat "timey" partitions differently, we'll need to have two separate conditionals, one to check if it's a regular TimeWindowPartitionsDefinition, and another to check if it's a MultiPartitionsDefinition with exactly one time dimension.

Once we have validated if we're in a "timey" case, then we often want to pass that "timey" partitions def into another function. In this case, we're doing that for `_sort_key_for_time_window_partition`, but we also do a similar thing in [this function](https://sourcegraph.com/github.com/dagster-io/dagster/-/blob/python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py?L165), for example. In each of these cases, we need to include extra logic to disambiguate between the TimeWindowPartitionsDefinition and MultiPartitionsDefinitions, as even if they share some properties, there are some parts which are different (e.g. in the sort key case, we have to pull out the partition key in a slightly different way, even if we get the `fmt` in the same way)

So this would all be a bit simpler if we could just have `_sort_key_for_time_window_partition` just take in a TimeWindowPartitionsDefinition object directly, regardless of if it came from a MultiPartitionsDef or not.

This also helps prevent mistakes where we accidentally pass in a MultiPartitionsDefinition that does not have exactly one time dimension, by letting us rely on the static type checker.

Reply:
Thanks for explaining.

I'll change the code around to make `_sort_key_for_time_window_partition` accept a time partitions def, though at this callsite (and other related callsites like the one you mentioned) there has to be a point where the code diverges to distinguish between time partitioned and multipartitioned keys.

Reply:
that's a good point about having to distinguish between time partition and multi partition keys -- in that case, could we have a `get_time_partition_key(partitions_def, partition_key)` to handle that (then pass in both the partitions_def and the time_partition_key into that function)?

my basic feeling is that if we want to have a function that handles something with time partitions, it'd be great to be able to only pass it things that are already in the language of time partitions, so that it can be completely concerned with handling time-based stuff